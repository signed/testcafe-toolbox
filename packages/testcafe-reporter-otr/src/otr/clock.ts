export interface Clock {
  now(): Date
}

export class SystemTime implements Clock {
  now(): Date {
    return new Date()
  }
}
