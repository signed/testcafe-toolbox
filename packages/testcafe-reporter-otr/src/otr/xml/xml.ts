import { createCB } from 'xmlbuilder2'

export type XMLBuilderCB = ReturnType<typeof createCB>

export type Context = {
  readonly xmlBuilder: XMLBuilderCB
  readonly namespaceRegistry: NamespaceRegistry
}

export const toXsDateTime = (date: Date) => date.toISOString()

export type XmlEmitter = (context: Context) => void

export class NamespaceRegistry {
  static of(defaultNamespace: Namespace, additionalNamespacesArg: Record<string, Namespace> = {}) {
    const additionalNamespaces = new Map()

    Object.entries(additionalNamespacesArg).forEach(([alias, namespace]) => {
      additionalNamespaces.set(namespace.uri, alias)
    })

    return new NamespaceRegistry(defaultNamespace, additionalNamespaces)
  }

  readonly defaultNamespace: Namespace
  readonly additionalNamespaces: Map<string, string>

  constructor(defaultNamespace: Namespace, additionalNamespaces: Map<string, string>) {
    this.defaultNamespace = defaultNamespace
    this.additionalNamespaces = additionalNamespaces
  }
}

type AttributeValue = string | number | Date

const attributeValueToString = (value: AttributeValue): string => {
  if (value instanceof Date) {
    return toXsDateTime(value)
  }
  if (typeof value === 'number') {
    return value.toString(10)
  }
  return value
}

export class Element {
  private readonly _context: Context
  private readonly _namespace: Namespace
  private readonly _simpleName: string

  constructor(context: Context, qualifiedName: QualifiedName) {
    this._context = context
    this._namespace = qualifiedName.namespace
    this._simpleName = qualifiedName.simpleName
  }

  openTag() {
    if (this._namespace.equals(this._context.namespaceRegistry.defaultNamespace)) {
      this._context.xmlBuilder.ele(this._simpleName)
    } else {
      this._context.xmlBuilder.ele(this._namespace.uri, this._simpleName)
    }
    return this
  }

  closeTag() {
    this._context.xmlBuilder.up()
    return this
  }

  withAttribute(qualifiedName: QualifiedName, value: AttributeValue) {
    let valueAsString = attributeValueToString(value)
    if (this._namespace.equals(qualifiedName.namespace)) {
      this._context.xmlBuilder.att(qualifiedName.simpleName, valueAsString)
    } else {
      this._context.xmlBuilder.att(qualifiedName.namespace.uri, qualifiedName.simpleName, valueAsString)
    }
    return this
  }

  withContent(version: string) {
    this._context.xmlBuilder.txt(version)
    return this
  }

  append(emitter: XmlEmitter): this {
    emitter(this._context)
    return this
  }
}

export class Namespace {
  static of(uri: string) {
    return new Namespace(uri)
  }

  readonly uri: string

  constructor(uri: string) {
    this.uri = uri
  }

  equals(other: Namespace) {
    return this.uri === other.uri
  }
}

export class QualifiedName {
  static of(namespace: Namespace, simpleName: string) {
    return new QualifiedName(namespace, simpleName)
  }

  readonly namespace: Namespace
  readonly simpleName: string

  constructor(namespace: Namespace, simpleName: string) {
    this.namespace = namespace
    this.simpleName = simpleName
  }
}
