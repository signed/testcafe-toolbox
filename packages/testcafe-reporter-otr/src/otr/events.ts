import { createWriteStream } from 'fs'
import { createCB } from 'xmlbuilder2'
import { Context, Element, Namespace, NamespaceRegistry, QualifiedName, XmlEmitter } from './xml'

export const eventsNamespace = Namespace.of('https://schemas.opentest4j.org/reporting/events/0.1.0')

const constructNamespacesFrom = (context: Context) => {
  const eventsAttributes: Record<string, string> = {
    xmlns: context.namespaceRegistry.defaultNamespace.uri,
  }

  context.namespaceRegistry.additionalNamespaces.forEach((prefix, namespaceUri) => {
    eventsAttributes['xmlns:' + prefix] = namespaceUri
  })
  return eventsAttributes
}

export interface Writer {
  write(chunk: string): void
  close(): void
}

export const intoFile = (filename: string): Writer => {
  const outStream = createWriteStream(filename)

  return new (class implements Writer {
    write(chunk: string): void {
      outStream.write(chunk)
    }
    close(): void {
      outStream.close()
    }
  })()
}

export const intoString = (): Writer & { readonly xml: string } => {
  let xml = ''
  return new (class implements Writer {
    get xml() {
      return xml
    }
    write(chunk: string): void {
      xml += chunk
    }
    close(): void {
      // noting to close
    }
  })()
}

export class EventsWriter {
  private readonly namespaceRegistry: NamespaceRegistry
  private context: Context | undefined

  constructor(namespaceRegistry: NamespaceRegistry) {
    this.namespaceRegistry = namespaceRegistry
  }

  startEmitting(target: Writer) {
    const xmlBuilder = createCB({
      data: target.write,
      end: target.close,
      prettyPrint: true,
    })
    this.context = { xmlBuilder, namespaceRegistry: this.namespaceRegistry }
    const eventsAttributes = constructNamespacesFrom(this.context)
    xmlBuilder.dec().ele(eventsNamespace.uri, 'e:events', eventsAttributes)
    return this
  }

  append(emitter: XmlEmitter): this {
    if (this.context) {
      emitter(this.context)
    }
    return this
  }

  close() {
    if (this.context === undefined) {
      return
    }
    this.context.xmlBuilder.end()
  }
}

export class Started extends Element {
  static Element = QualifiedName.of(eventsNamespace, 'started')
  static Id = QualifiedName.of(eventsNamespace, 'id')
  static Name = QualifiedName.of(eventsNamespace, 'name')
  static Time = QualifiedName.of(eventsNamespace, 'time')
  static ParentId = QualifiedName.of(eventsNamespace, 'parentId')

  constructor(context: Context) {
    super(context, Started.Element)
  }

  withParentId(parentId: string) {
    this.withAttribute(Started.ParentId, parentId)
    return this
  }
}

export const started = (id: string, name: string, time: Date, consumer?: (started: Started) => void) => {
  return (context: Context) => {
    const started = new Started(context).openTag()
    started.withAttribute(Started.Id, id)
    started.withAttribute(Started.Name, name)
    started.withAttribute(Started.Time, time)
    consumer?.(started)
    started.closeTag()
  }
}

export class Finished extends Element {
  static Element = QualifiedName.of(eventsNamespace, 'finished')
  static Id = QualifiedName.of(eventsNamespace, 'id')
  static Time = QualifiedName.of(eventsNamespace, 'time')

  constructor(context: Context) {
    super(context, Finished.Element)
  }

  withId(id: string) {
    this.withAttribute(Finished.Id, id)
    return this
  }

  withTime(time: Date) {
    this.withAttribute(Finished.Time, time)
    return this
  }
}

export const finished = (id: string, time: Date, consumer?: (finished: Finished) => void) => {
  return (context: Context) => {
    const finished = new Finished(context).openTag()
    finished.withId(id)
    finished.withTime(time)
    consumer?.(finished)
    finished.closeTag()
  }
}
