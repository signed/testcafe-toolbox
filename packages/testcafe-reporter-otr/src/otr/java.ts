import { Context, Element, Namespace, QualifiedName } from './xml/xml'

export const javaNamespace = Namespace.of('https://schemas.opentest4j.org/reporting/java/0.1.0')

class JavaVersion extends Element {
  static Element = QualifiedName.of(javaNamespace, 'javaVersion')

  constructor(context: Context) {
    super(context, JavaVersion.Element)
  }
}

export const javaVersion = (version: string) => {
  return (context: Context) => {
    const javaVersion = new JavaVersion(context).openTag()
    javaVersion.withContent(version)
    javaVersion.closeTag()
  }
}
