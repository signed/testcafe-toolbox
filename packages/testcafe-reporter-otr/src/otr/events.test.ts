import { expect, test } from '@jest/globals'
import {
  directorySource,
  filePosition,
  fileSource,
  hostName,
  result,
  userName,
  infrastructure,
  coreNamespace,
} from './core'
import { eventsNamespace, EventsWriter, finished, intoString, started } from './events'
import { javaNamespace, javaVersion } from './java'
import { NamespaceRegistry } from './xml/xml'
import { Clock } from './clock'

test('events example', () => {
  const clock = new FixedTime(new Date(1664127808347))

  const namespaceRegistry = NamespaceRegistry.of(coreNamespace, { e: eventsNamespace, java: javaNamespace })
  const target = intoString()
  const events = new EventsWriter(namespaceRegistry).startEmitting(target)

  // prettier-ignore
  events.append(infrastructure((_) => _
            .append(hostName('wonderland'))
            .append(userName('alice'))
            .append(javaVersion('1.8'))
        )
    );
  events.append(started('1', 'container', clock.now()))
  events.append(started('2', 'test', clock.now(), (started) => started.withParentId('1')))
  events.append(finished('2', clock.now(), (finished) => finished.append(result('FAILED'))))
  events.append(
    finished('1', clock.now(), (finished) =>
      finished
        .append(directorySource('/tmp/directory'))
        .append(
          fileSource('/tmp/screenshot.png', (fileSource) => {
            fileSource.append(filePosition(42, 17))
          }),
        )
        .append(result('FAILED')),
    ),
  )
  events.close()
  expect(target.xml).toMatchSnapshot()
})

export class FixedTime implements Clock {
  private readonly _now: Date

  constructor(now: Date) {
    this._now = now
  }

  now() {
    return this._now
  }
}
