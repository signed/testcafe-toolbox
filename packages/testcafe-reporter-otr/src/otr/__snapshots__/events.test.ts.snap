// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`events example 1`] = `
"<?xml version="1.0"?>
<e:events xmlns="https://schemas.opentest4j.org/reporting/core/0.1.0" xmlns:e="https://schemas.opentest4j.org/reporting/events/0.1.0" xmlns:java="https://schemas.opentest4j.org/reporting/java/0.1.0">
  <infrastructure>
    <hostName>
      wonderland
    </hostName>
    <userName>
      alice
    </userName>
    <java:javaVersion>
      1.8
    </java:javaVersion>
  </infrastructure>
  <e:started id="1" name="container" time="2022-09-25T17:43:28.347Z"/>
  <e:started id="2" name="test" time="2022-09-25T17:43:28.347Z" parentId="1"/>
  <e:finished id="2" time="2022-09-25T17:43:28.347Z">
    <result status="FAILED"/>
  </e:finished>
  <e:finished id="1" time="2022-09-25T17:43:28.347Z">
    <directorySource path="/tmp/directory"/>
    <fileSource path="/tmp/screenshot.png">
      <filePosition line="42" column="17"/>
    </fileSource>
    <result status="FAILED"/>
  </e:finished>
</e:events>"
`;
